pkg/controller/support/credentials/cloudauth/acr/acr_test.go

```diff
@@ -0,0 +1,163 @@
+			nil,
+			fmt.Sprintf(`ACR URL is invalid: "test-server" should match pattern %s`, acrRegex),
+			invalidServerErr,
+		},
+		{
+			"failed_get_from_exchange",
+			"foo.azurecr.us",
+			createProvider("test-tenantId", fakeServicePrincipalToken{}),
+			createDefaultRefreshTokensClient(true),
+			cloudauth.CreateDefaultChallengeLoginServer("", "", nil),
+			nil,
+			"Token refresh failed.",
+			acrGetExchangeErr,
+		},
+		{
+			"failed_refresh_exchange",
+			"foo.azurecr.us",
+			createProvider("test-tenantId", fakeServicePrincipalToken{errOut: true}),
+			createDefaultRefreshTokensClient(false),
+			cloudauth.CreateDefaultChallengeLoginServer("", "", nil),
+			nil,
+			"Failed to refresh AAD token.",
+			aadTokenErr,
+		},
+		{
+			"failed_default_challenge_login_server",
+			"foo.azurecr.us",
+			createProvider("test-tenantId", fakeServicePrincipalToken{}),
+			createDefaultRefreshTokensClient(false),
+			cloudauth.CreateDefaultChallengeLoginServer("", "", challengeServerErr),
+			nil,
+			"ACR cloud authentication failed.",
+			aadTokenErr,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			observerCore, observedLogs := observer.New(zap.DebugLevel)
+			log := zapr.NewLogger(zap.New(observerCore))
+
+			defaultChallengeLoginServer = tt.defaultChallengeLoginServer
+			defaultRefreshTokensClient = tt.defaultRefreshTokensClient
+
+			authConfig, err := tt.provider.authenticate(ctx, log, tt.serverName)
+			assert.Equal(t, tt.authConfig, authConfig)
+
+			// Compare expected error condition.
+			if tt.expectedError == nil {
+				require.NoError(t, err)
+			} else {
+				require.Error(t, err)
+				assert.Equal(t, tt.expectedError.Error(), err.Error())
+			}
+
+			// Compare expected log messages.
+			logLen := observedLogs.Len()
+			require.GreaterOrEqual(t, logLen, 1)
+			recentLogMessage := observedLogs.All()[observedLogs.Len()-1].Message
+			assert.Equal(t, tt.expectedLogMessage, recentLogMessage)
+		})
+	}
+}
+
+type fakeServicePrincipalToken struct {
+	errOut bool
+}
+
+func (f fakeServicePrincipalToken) RefreshWithContext(ctx context.Context) error { return nil }
+
+func (f fakeServicePrincipalToken) RefreshExchangeWithContext(ctx context.Context, resource string) error {
+	return nil
+}
+
+func (f fakeServicePrincipalToken) EnsureFreshWithContext(ctx context.Context) error {
+	if f.errOut {
+		return errors.New("failed to refresh principal token")
+	}
+
+	return nil
+}
+
+func (f fakeServicePrincipalToken) OAuthToken() string {
+	return "test-oauth-token"
+}
+
+type fakeRefreshTokensClient struct {
+	errOut bool
+}
+
+func (f fakeRefreshTokensClient) GetFromExchange(_ context.Context, _, _, _, _, _ string) (result containerregistry.RefreshToken, err error) {
+	if f.errOut {
+		err = errors.New("get from exchange error")
+	}
+	return
+}
+
+// Helper functions
+func createProvider(tenantId string, servicePrincipalToken fakeServicePrincipalToken) *acrProvider {
```

Comment:
I think you can simplify the arguments here so that it takes `(tenantID string, spTokenErr bool)`, right?